### ISSUE TITLE
Note when opaque future from `async fn` is involved in a type mismatch is unclear

### ISSUE THREAD

**Issue by SNCPlay42** [rust-lang/rust#80658]
```rust
fn foo() -> u8 {
    async fn async_fn() -> () {}
    
    async_fn()
}
```
emits
```
error[E0308]: mismatched types
 --> src/lib.rs:4:5
  |
1 | fn foo() -> u8 {
  |             -- expected `u8` because of return type
2 |     async fn async_fn() -> () {}
  |                            -- the `Output` of this `async fn`'s found opaque type
3 |
4 |     async_fn()
  |     ^^^^^^^^^^ expected `u8`, found opaque type
  |
  = note:     expected type `u8`
          found opaque type `impl Future`
```

The note "the `Output` of this `async fn`'s found opaque type" is kind of hard to understand. If the return type of `async_fn` isn't explicitly given, it doesn't even point at a type:
```
2 |     async fn async_fn() {}
  |                         - the `Output` of this `async fn`'s found opaque type
```

Perhaps the note should instead look like this:

```
2 |     async fn async_fn() {}
  |     ----- the found opaque type is the `impl Future` returned by this `async fn`
```

@rustbot label A-async-await A-diagnostics D-confusing T-compiler

**Comment by nellshamrell** [comment]
Looks like #66463 touched areas of code that are relevant to this issue (not to say that it caused the issue, only that it provides a potential pointer to the parts of the code related to generating this error message)

**Comment by tmandry** [comment]
@nellshamrell Are you still planning to work on this? If not no worries, feel free to release your assignment.

**Comment by henryboisdequin** [comment]
Can't this issue be closed (#82165) or is there still more work to be done?

**Comment by camelid** [comment]
The note seems clearer now:

```
error[E0308]: mismatched types
 --> src/lib.rs:4:5
  |
1 | fn foo() -> u8 {
  |             -- expected `u8` because of return type
2 |     async fn async_fn() -> () {}
  |                            -- checked the `Output` of this `async fn`, found opaque type
3 |     
4 |     async_fn()
  |     ^^^^^^^^^^ expected `u8`, found opaque type
  |
  = note: while checking the return type of the `async fn`
  = note:     expected type `u8`
          found opaque type `impl Future`
```

But it would be good to get confirmation that this is fixed before closing.

**Comment by SNCPlay42** [comment]
"checked the `Output` of this `async fn`, found opaque type" still seems confusing to me; that makes it sound like the mismatch is in the `Output` of the `Future` returned by the function, which isn't the case. The mismatch is that it *is* a `Future<Output = T>`, regardless of what `T` is, instead of just a `u8`.

(Oops, did I confuse the issue by giving `foo` and `async_fn` different nominal return types? It would still be a mismatch if `async_fn` was declared as `async fn async_fn() -> u8`.)

**Comment by nikomatsakis** [comment]
I'd prefer to see a more radically improved error message. What I posted on #84166 was

```
error[E0308]: missing await on future
 --> src/main.rs:8:12
  |
3 | async fn return_number() -> usize {
  | ----- calling an async function returns a future
...
8 |     return return_number();
  |            ^^^^^^^^^^^^^^^ expected `usize`, found future
  |
help: consider `await`ing on the `Future`
  |
8 |     return return_number().await;
  |                           ^^^^^^

error: aborting due to previous error
```

**Comment by nikomatsakis** [comment]
So I was looking at this in order to write mentoring instructions. I note that the example in the OP contains a kind of type error (the async function is returning `()`, but the outer function is returning `u8`) -- @SNCPlay42 was that intentional?

If we rewrite to this example:

```rust
fn foo() -> u8 {
    async fn async_fn() -> u8 {  22 }
    
    async_fn()
}
```

we presently get this error:

```
error[E0308]: mismatched types
 --> src/lib.rs:4:5
  |
1 | fn foo() -> u8 {
  |             -- expected `u8` because of return type
2 |     async fn async_fn() -> u8 {  22 }
  |                            -- the `Output` of this `async fn`'s found opaque type
3 |     
4 |     async_fn()
  |     ^^^^^^^^^^ expected `u8`, found opaque type
  |
  = note:     expected type `u8`
          found opaque type `impl Future`
help: consider `await`ing on the `Future`
  |
4 |     async_fn().await
  |               ^^^^^^
```

**Comment by nikomatsakis** [comment]
I would like to give this error;

```
error[E0308]: missing await on future
 --> src/main.rs:8:12
  |
1 | fn foo() -> u8 {
  |             -- expected `u8` because of return type
2 |     async fn async_fn() -> u8 {  22 }
  |     ----- calling an async function returns a future
3 |     
4 |     async_fn()
  |     ^^^^^^^^^^ expected `u8`, found future
  |
help: consider `await`ing on the `Future`
  |
4 |     async_fn().await
  |               ^^^^^^

error: aborting due to previous error
```

which isn't that far apart.

**Comment by nikomatsakis** [comment]
I believe the current "help" is generated by this function:

https://github.com/rust-lang/rust/blob/e5615dcc58119c401dd4824fee2e3052d1b2d9be/compiler/rustc_infer/src/infer/error_reporting/mod.rs#L1731-L1755

although there are some other places that generate similar messages (e.g., [here](https://github.com/rust-lang/rust/blob/e5615dcc58119c401dd4824fee2e3052d1b2d9be/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs#L2265-L2271)).

**Comment by nikomatsakis** [comment]
This in turn is invoked from 

https://github.com/rust-lang/rust/blob/e5615dcc58119c401dd4824fee2e3052d1b2d9be/compiler/rustc_infer/src/infer/error_reporting/mod.rs#L1424-L1431

**Comment by nikomatsakis** [comment]
This could be a bit difficult because the setup is currently very oriented at adding helps and notes to an existing error (`diag`). To generate the error the way I wanted it would require "resetting' `diag` to a whole different error pattern, probably by detecting the situation here instead:

https://github.com/rust-lang/rust/blob/e5615dcc58119c401dd4824fee2e3052d1b2d9be/compiler/rustc_infer/src/infer/error_reporting/mod.rs#L1941-L1946

although `note_type_err` is also called from other places, like [here](https://github.com/rust-lang/rust/blob/e5615dcc58119c401dd4824fee2e3052d1b2d9be/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs#L1263-L1264).

**Comment by nikomatsakis** [comment]
You'll have to dig a bit through the code to see just how hard it is to decide when the `await` message applies, although you can see from the OP that even if the types aren't exactly equal, there's probably a better error to be given when one type is a future and the other is not. 

I think if I were going to do this I might keep the logic in `note_type_err`, but also have `report_and_explain_type_error` discover the situation and report a different sort of error, although it'd be good to make some examples where `note_type_err`' logic still triggers.

**Comment by nikomatsakis** [comment]
Not sure @nellshamrell if that helps at all =)

**Comment by nellshamrell** [comment]
FINALLY have some time to devote to this. Thank you so much for the pointers, @nikomatsakis, that helps a lot!

**Comment by nellshamrell** [comment]
Adding some repro steps for my own reference (and for anyone else!)

```bash
$ cargo new troubleshooting_example
$ cd troubleshooting_example
```

Head into `src/lib.rs` and change to:

```rust
fn foo() -> u8 {
    async fn async_fn() -> u8 {  22 }
    
    async_fn()
}
```

```bash
$ cargo build

error[E0308]: mismatched types
 --> src/lib.rs:4:5
  |
1 | fn foo() -> u8 {
  |             -- expected `u8` because of return type
2 |     async fn async_fn() -> u8 {  22 }
  |                            -- the `Output` of this `async fn`'s found opaque type
3 |     
4 |     async_fn()
  |     ^^^^^^^^^^ expected `u8`, found opaque type
  |
  = note:     expected type `u8`
          found opaque type `impl std::future::Future`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `troubleshooting_example`.

To learn more, run the command again with --verbose.
```

**Comment by nellshamrell** [comment]
# Notes

## Relevant files/functions

### compiler/rustc_infer/src/infer/error_reporting/mod.rs

**add_labels_for_types**

This function adds some wording async errors (this is where I focused in #82165)

```rust
            fn add_labels_for_types(
                &self,
                err: &mut DiagnosticBuilder<'_>,
                target: &str,
                types: &FxHashMap<TyCategory, FxHashSet<Span>>,
            ) {
                for (key, values) in types.iter() {
                    let count = values.len();
                    let kind = key.descr();
                    let mut returned_async_output_error = false;
                    for sp in values {
                        err.span_label(
                            *sp,
                            format!(
                                "{}{}{} {}{}",
                                if sp.is_desugaring(DesugaringKind::Async)
                                    && !returned_async_output_error
                                {
                                    "checked the `Output` of this `async fn`, "
                                } else if count == 1 {
                                    "the "
                                } else {
                                    ""
                                },
                                if count > 1 { "one of the " } else { "" },
                                target,
                                kind,
                                pluralize!(count),
                            ),
                        );
                        if sp.is_desugaring(DesugaringKind::Async)
                            && returned_async_output_error == false
                        {
                            err.note("while checking the return type of the `async fn`");
                            returned_async_output_error = true;
                        }
                    }
                }
            }
        }
```

**suggest_await_on_expect_found**

This appears to generate the help text in the relevant async error from the example, particularly this section:

```rust
            (_, Some(ty)) if ty::TyS::same_type(exp_found.expected, ty) => {
                let span = match cause.code {
                    // scrutinee's span
                    ObligationCauseCode::Pattern { span: Some(span), .. } => span,
                    _ => exp_span,
                };
                diag.span_suggestion_verbose(
                    span.shrink_to_hi(),
                    "consider `await`ing on the `Future`",
                    ".await".to_string(),
                    Applicability::MaybeIncorrect,
                );
            }
```

**note_type_err**

This function calls `suggest_await_on_expect_found`

```rust
        if let Some(exp_found) = exp_found {
            self.suggest_as_ref_where_appropriate(span, &exp_found, diag);
            self.suggest_accessing_field_where_appropriate(cause, &exp_found, diag);
            self.suggest_await_on_expect_found(cause, span, &exp_found, diag);
        }
```

**Comment by nikomatsakis** [comment]
@nellshamrell sounds good-- let me know if I can help

**Comment by nellshamrell** [comment]
I've completed a partial fix for this at this time.

When running the above example, I now get this output:

```bash
nell@DESKTOP-VBH9MU6:~/rust/troubleshooting_example$ rustc +rust-stage-1 src/lib.rs --edition 2018
error[E0601]: `main` function not found in crate `lib`
 --> src/lib.rs:1:1
  |
1 | / fn foo() -> u8 {
2 | |     async fn async_fn() -> u8 {  22 }
3 | |
4 | |     async_fn()
5 | | }
  | |_^ consider adding a `main` function to `src/lib.rs`

error[E0308]: mismatched types
 --> src/lib.rs:4:5
  |
1 | fn foo() -> u8 {
  |             -- expected `u8` because of return type
2 |     async fn async_fn() -> u8 {  22 }
  |                            -- calling an async function returns a future
3 |
4 |     async_fn()
  |     ^^^^^^^^^^ expected `u8`, found opaque type
  |
  = note: while checking the return type of the `async fn`
  = note:     expected type `u8`
          found opaque type `impl Future`
help: consider `await`ing on the `Future`
  |
4 |     async_fn().await
  |               ^^^^^^

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0308, E0601.
For more information about an error, try `rustc --explain E0308`.
```

Seeing if I can further improve it.

**Comment by nellshamrell** [comment]
I'm trying to figure out where in the code "^^^^^^^^^^ expected `u8`, found opaque type" is generated.

**Comment by nikomatsakis** [comment]
@nellshamrell there is some generic machinery to manage type mismatches, I guess it's there *somewhere*

**Comment by nellshamrell** [comment]
I've opened a PR with at least a partial fix for this issue:  #86705

**Comment by eholk** [comment]
Is this still an issue?

The current output for the test case ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0c7aa82879ff0f48f11b4f28f5a61497)) is:
```
Compiling playground v0.0.1 (/playground)
error[[E0308]](https://doc.rust-lang.org/stable/error-index.html#E0308): mismatched types
 --> src/lib.rs:4:5
  |
1 | fn foo() -> u8 {
  |             -- expected `u8` because of return type
...
4 |     async_fn()
  |     ^^^^^^^^^^ expected `u8`, found opaque type
  |
note: while checking the return type of the `async fn`
 --> src/lib.rs:2:28
  |
2 |     async fn async_fn() -> u8 {  22 }
  |                            ^^ checked the `Output` of this `async fn`, found opaque type
  = note:     expected type `u8`
          found opaque type `impl Future<Output = u8>`
help: consider `await`ing on the `Future`
  |
4 |     async_fn().await
  |               ++++++

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` due to previous error
```
which looks pretty close to the desired output listed above.
